segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -484,19 +487,44 @@ private EventsReadAndTruncationLength deserializeEvents(BufferView inputData) {
+         * @return CompletableFuture that, when completed successfully, indicates that the provided handlers has been
+         *         executed on the events read from the internal Segment.
+         */
         private CompletableFuture<EventsReadAndTruncationLength> applyProcessorHandler(EventsReadAndTruncationLength readResult) {
             return this.handler.apply(readResult.getProcessorEventsData()).thenApply(v -> readResult);
         }
 
         private CompletableFuture<Void> truncateInternalSegment(EventsReadAndTruncationLength readResult, Timer iterationTime) {
-            long truncationOffset = this.segmentStartOffset.get() + readResult.getTruncationLength();
+            Preconditions.checkState(this.lastTruncationOffset.get() >= 0);
+            Preconditions.checkState(this.lastTruncationOffset.get() <= this.processedUpToOffset.get());
+            // Update the processedUpToOffset to the current offset that has been processed.
+            this.processedUpToOffset.addAndGet(readResult.getTruncationLength());
+            // Only do the actual truncation if the internal Segment has accumulated the configured amount of data.
+            // Note that a restart will induce re-processing of the non-truncated tasks, but this is fine as this class
+            // ensured at-lest-one processing guarantees.
+            CompletableFuture<Void> tryTruncate = shouldTruncate() ? doTruncateInternalSegment() : CompletableFuture.completedFuture(null);
+            return tryTruncate.thenAccept(v -> {
+                                // Report latency metrics upon complete processing iteration (irrespective of if truncation happened or not).
+                                this.metrics.batchProcessingLatency(iterationTime.getElapsedMillis());
+                            });
+        }
+
+        private boolean shouldTruncate() {
+            return this.processedUpToOffset.get() - this.lastTruncationOffset.get() >= this.config.getProcessedDataTruncationSizeInBytes();
+        }
+
+        private CompletableFuture<Void> doTruncateInternalSegment() {
+            final long truncationOffset = this.processedUpToOffset.get();
+            Preconditions.checkState(truncationOffset <= this.segment.getInfo().getLength());
             return this.segment.truncate(truncationOffset, this.containerOperationTimeout)
-                               .thenAccept(v -> {
-                                   // Update the startSegmentOffset after truncation.
-                                   this.segmentStartOffset.addAndGet(readResult.getTruncationLength());
-                                   // Report latency metrics upon complete processing iteration.
-                                   this.metrics.batchProcessingLatency(iterationTime.getElapsedMillis());
-                               });
+                    .thenAccept(v -> {
+                        // Reset the truncation offset to the new one.
```

Comment:
would a metric indicating truncated or processed bytes so far be indicative of anything?

Reply:
In the case of error, we already output the start offset of the segment and the processing offset, but there are no metrics for these. I don't think that we need a metric for this neither.